pkg/proxy/nftables/proxier.go

```diff
@@ -1015,6 +1045,104 @@ func isAffinitySetName(set string) bool {
 	return strings.HasPrefix(set, servicePortEndpointAffinityNamePrefix)
 }
 
+// nftElementStorage is an internal representation of nftables map with single string value or set.
+// Currently only supports sets and maps with single string value.
+type nftElementStorage struct {
+	elements      map[string]string
+	leftoverKeys  sets.Set[string]
+	containerType string
+	containerName string
+}
+
+func (c *nftElementStorage) joinKey(k []string) string {
+	return strings.Join(k, " . ")
+}
+
+func (c *nftElementStorage) splitKey(k string) []string {
+	return strings.Split(k, " . ")
+}
+
+// newNFTElementStorage creates an empty nftElementStorage.
+// nftElementStorage.reset() must be called before the first usage.
+func newNFTElementStorage(containerType, containerName string) *nftElementStorage {
+	c := &nftElementStorage{
+		elements:      make(map[string]string),
+		leftoverKeys:  sets.New[string](),
+		containerType: containerType,
+		containerName: containerName,
+	}
+	return c
+}
+
+// reset clears the internal state and flushes the nftables map/set.
+func (c *nftElementStorage) reset(tx *knftables.Transaction) {
+	clear(c.elements)
+	if c.containerType == "set" {
+		tx.Flush(&knftables.Set{
+			Name: c.containerName,
+		})
+	} else {
+		tx.Flush(&knftables.Map{
+			Name: c.containerName,
+		})
+	}
+}
+
+// resetLeftoverKeys should be called before every transaction to ensure cleanupLeftoverKeys
+// will have correct data.
+func (c *nftElementStorage) resetLeftoverKeys() {
+	clear(c.leftoverKeys)
+	for key := range c.elements {
+		c.leftoverKeys.Insert(key)
+	}
+}
+
+// getValue converts element value to string.
+// It returns empty string for sets, and the first string for maps.
+// If a map value has 0 elements, it returns an empty string, but the following transaction will fail.
+func (c *nftElementStorage) getValue(v []string) string {
+	if c.containerType == "set" || len(v) == 0 {
+		return ""
+	} else {
+		return v[0]
+	}
+}
+
+// ensureElem adds elem to the transaction if elem is not present in the container, and updates internal
+// leftoverKeys set to track unused elements.
+func (c *nftElementStorage) ensureElem(tx *knftables.Transaction, elem *knftables.Element) {
+	newKey := c.joinKey(elem.Key)
+	newValue := c.getValue(elem.Value)
+	existingValue, ok := c.elements[newKey]
+	if ok {
```

Comment:
IMO, "`ok`" is fine to use as the optional second variable name for a typecheck or channel read, where one of the two outcomes is clearly "not ok", but you shouldn't do it for a map lookup (since the map value not existing is _not_ an error). Use `existingValue, exists := c.elements[newKey]`.